SJRK-104: Use pushState to manage back/forward behaviour when creating a story

BlueSlug opened the pull request:
This update implements the task as described in [the Jira](https://issues.fluidproject.org/browse/SJRK-104). I made the choice not to include the state variables in the query string, but this is easily enabled by removing the exclusion rules in the `historian` component.

This update also addresses [SJRK-261](https://issues.fluidproject.org/browse/SJRK-261).

cindyli committed:
Merge remote-tracking branch 'gregor/SJRK-255' into stories-floe-dev

* gregor/SJRK-255:
  SJRK-255: re-alphabetized dependencies
  SJRK-283: solving FLUID-6394 has solved this issue as well
  SJRK-255: narrowed down and tidied overly-specific options distributions
  SJRK-255: removed unused fields and things commented out for SJRK-132 See the SJRK-132 for more information: https://issues.fluidproject.org/browse/SJRK-132?focusedCommentId=40931
  SJRK-255: solved race condition in UIO Table of Contents modelListener
  NOJIRA: made test event specification less pointy
  SJRK-255: simplifying IoCSS to avoid distributing to option of option of ...and so on
  SJRK-255: updated gpii-binder, removed workaround for GPII-4259 bug
  SJRK-255: added resetPreferences for page, updated tests, removed unused
  SJRK-255: updated infusion, fixed Edit and Browse page tests
  SJRK-255: WIP refactored page/uio interaction, updated tests
  SJRK-255: added namespace to options dist. record to avoid conflicts
  SJRK-255: updated infusion to latest dev release
  SJRK-255: updated fluid-couch-config to use my fork instead of waharnum
  SJRK-255: WIP continued tidying up UIOMultilingualDemo.js
  SJRK-225: WIP reduced duplication in UIO grade options distrubution It's still broken though.
  SJRK-255: WIP un-hid the language changing links in the base theme
  SJRK-255: WIP commit of trying to get UIO working with immutable options - moved multilingualSettings values to model - removed `reloadUioMessages` function and calls - gave names to some options distribution entries - got TOC enactor to load **sometimes**, there is an as-yet unsolved race condition - removed local copies of UIO l10n messages as they're no longer necessary
  SJRK-255: added note for GPII-4259 and updated PouchDB to solve errors
  SJRK-255: docs update for clarity
  SJRK-255: temp fix for gpii.binder mergePolicy bug
  SJRK-255: updated dependencies and removed mutation of l10n `options`
  SJRK-299: removed contentString and related bits, as well as other cruft cruft includes: - languageFromSelect and languageFromInput values - contentString model relays - tests for contentString values and test files which consisted entirely of such tests - the "restore block" button and css rules - unused images - unnecesary values for the example story - localized messages which are no longer used - anything else I may have missed
  SJRK-255: moved listeners to model relay with publishing state mapping
  SJRK-255: converted progress and response areas to model-based logic
  SJRK-255: removed obsolete logic for discarded save-no-share button
  SJRK-255: converted share button logic to model/listener
  SJRK-255: tidying things up a little, reducing number of cookie calls
  SJRK-255: moved previewer component to its own grade
  NOJIRA: re-adding missing package from `upstream/stories-floe-dev`
  SJRK-255: added missing JS file references to non-test HTML files
  SJRK-255: updated Infusion to latest FLUID-6148 dev release
  SJRK-255: updated secrets.json instructions for a bit of clarity
  SJRK-255: WIP - updated test files to conform with package updates
  Work in progress updating to FLUID-6148 branch. Server tests run, starting work on client tests which need to be fixed up for include structure of more recent gpii-handlebars release
  SJRK-255: loadBrowse tests updated to reflect changes since introduction
  SJRK-255: re-add breaking code to the project. requires db server to run

BlueSlug committed:
SJRK-104: editor "pages" -> "parts" to differentiate from page grade

BlueSlug committed:
SJRK-104: invokerized visibility management calls for less coupling
more to come :)

BlueSlug committed:
SJRK-104: added gpii-location-bar-relay component to manage page state
I named it "historian" after the near-identical grade I began to write 
before discovering this one already exists

BlueSlug committed:
SJRK-267: consolidated copyright dates in new AUTHORS.md, added refs

BlueSlug committed:
SJRK-104: "parts" -> "steps", updated step names to be more descriptive

BlueSlug committed:
SJRK-104: started converting storyEditor to use visibility model state

BlueSlug committed:
Merge remote-tracking branch 'upstream/master' into SJRK-104

BlueSlug committed:
SJRK-104: implemented browser state management, removed old events setup

BlueSlug commented:
Please hold off reviewing for the time being, I will be adding tests specific to the `historian` component and solving some sequence breaks that existed before this work

BlueSlug committed:
SJRK-261: removed extraneous `page` listener/event in `previewer` grade
Also moved some event wiring to use the new model values instead. This 
effectively deals with SJRK-261.

BlueSlug committed:
SJRK-104: added extra steps to ensure no sequence breaks within editor

BlueSlug commented:
Ready for review.

I tried to add tests for the forward & back functionality, but I was getting browser security related errors ("TypeError: Illegal invocation") when trying to trigger those actions programatically.

BlueSlug committed:
SJRK-104: docs update to reflect new changes

BlueSlug committed:
Revert "SJRK-267: consolidated copyright dates in new AUTHORS.md, added refs"

This reverts commit a9480a9af9c1f25f7d13e450905281e20ae168af.

cindyli commented on src/ui/base-page-storyEdit.js:
```diff
@@ -62,6 +93,14 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
                 funcName: "sjrk.storyTelling.base.page.storyEdit.setEditorDisplay",
                 args: ["{that}.options.selectors.mainContainer", "{that}.options.selectors.pageContainer", "{that}.options.pageSetup.authoringEnabled", "{that}.options.pageSetup.hiddenEditorClass"]
+            },
+            showEditorHidePreviewer: {
```
`showEditorHidePreviewer()` and `hideEditorShowPreviewer()` invokers can be combined into one by accepting a boolean argument.

cindyli commented on src/ui/ui-storyEditor.js:
```diff
@@ -14,14 +14,33 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+            {
+                this: "{that}.dom.storyMetadataStep",
+                method: "toggle",
+                args: ["{change}.oldValue"],
```
I understand you use `{change}.oldValue` to retrieve the opposite of the current boolean value in `{change}.value`. However, it introduces a logic confusion that this action is controlled by the old model value rather than the current. Personally, I prefer to sacrifice a little coding convenience for clarifying the logic.

BlueSlug replied:
I agree that this method is confusing. What I'd like is to have a way to say "the opposite of `{change}.value`", but without the `negate` transform I was beginning to create (my work on that can be found in [my WIP branch](https://github.com/BlueSlug/sjrk-story-telling/blob/fd1f8096dcbf1ddd98a484c7dacc130412d898bb/src/ui/transforms.js#L111-L126)), I'm not sure how best to approach this.

What do you think of adding a basic "not" function locally in this project, with the aim of using it in an expander and passing in `{change}.value`? This function can later be supplanted by the `negate` transform, whenever that ends up being incorporated into Infusion.

BlueSlug replied:
I've just noticed @amb26's comment below suggesting to do something very similar

cindyli commented on src/ui/ui-storyEditor.js:
```diff
@@ -96,25 +115,27 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
                 funcName: "sjrk.storyTelling.ui.storyEditor.removeSelectedBlocks",
                 args: ["{that}", "{that}.blockManager.managedViewComponentRegistry"]
+            },
+            "onEditorNextRequested": {
```
For consistency, include the namespace with the event as above. 

The same comment applies to the listener at line 126.

cindyli commented:
`editorVisible` and `editStoryStepVisible` work together to control 3 steps of editing and previewing. It would be nice to add code comment or documentation to explain model values of these 2 at different steps.

amb26 commented on src/ui/base-page-storyEdit.js:
```diff
@@ -16,6 +16,40 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+            // in much the same way as within the editor, the visibility
+            // of the editor and previewer are mutually exclusive and
+            // represented by a single model value, for the time being
+            editorVisible: true
```
As below, with the use of "oldValue" which @cindyli has noted below. Rather than applying this historyful gimmick, you should have a properly organised model with separate flags for "editorVisible" and "previewerVisible". Connect them with a relay - presumably this is the situation which gave rise to the legendary standup update "Infusion doesn't have an inversion relay" : P
Looking at the impl for the fluid.transforms.linearScale transform which could have been hacked into service, I see that @kaspermarkus unhelpfully put some type checks into it, so I suggest you just apply
````
    sjrk.storyTelling.not = function (value) {
          return !value;
     };
````
and specify this as the relay rule.

amb26 replied:
For extra helpfulness, please contribute this to Infusion as fluid.transforms.not together with its self-inverse definition and a couple of small test cases :)

BlueSlug replied:
Indeed, this is precisely the situation that gave rise to my comments ;) I began to work on such a transform, but we decided that for the sake of expedience I would shelve that work and Just Get It Working™ with this binary view-control method, and come back to this when there is less pressing work to be done.

I've achieved a less confusing solution via [the method I proposed a couple review items above](https://github.com/fluid-project/sjrk-story-telling/pull/56#discussion_r377315295). It's a bit of a middle ground between using `{change}.oldValue` and re-implementing the model relay I once had (with either a free transform calling this new `not` function or as a full-fledged inversion transform). Not ideal, but perhaps good enough for now? I can file a Jira to make sure we repay this technical debt.

BlueSlug committed:
SJRK-104: consolidated four invokers into two for more DRYness
- hideEditorShowPreviewer and showEditorHidePreviewer combined in the 
latter and now accepts one boolean argument to control hiding and 
showing
- ditto for hideEditStoryStep and showEditStoryStep

BlueSlug committed:
SJRK-104: added a basic `not` function, using that for less confusion

BlueSlug committed:
SJRK-104: moving namespace into listener name for tidiness

BlueSlug committed:
SJRK-104: added more comments to clarify the hide/show steps mechanism

BlueSlug commented:
After discussing the one value / two value situation with @cindyli and @amb26, we have two possible solutions to the matter at hand, each with its pros and cons:

1. we continue using a single model value to represent the visibility states (being the steps of the editor or the `storyEditor` & `storyPreviewer` within the Edit page), and give it a better name to more clearly convey that it controls more than one thing
2. we revert to using multiple model values to control each piece individually, and we add model relays to coordinate them

Approach 1 sacrifices the granularity of individual control values for simplicity of implementation. We don't currently have a need for more than two states per value, so the only programming change necessary with this approach is to choose a better name. When requirements change, the single value can be adapted to use string values instead of boolean, in much the same way as the model relay and listeners in `storyPreviewer`. Downsides include tigher coupling and a steeper learning curve for any new developers.

Approach 2 has the benefit of clarity from one model value representing one thing (i.e. the visibility of one discrete element of the UI) as well as more loosely coupled code to make future development less complicated. The downside is, of course, that there will need to be more wiring (model relays, listeners and transforms) to maintain the relationships between the various values in order to change them with one mutator call, or more mutator calls to change all the relevant values. Reintroducing those bits of wiring would take more time, and they may be more repetitive.

When asked in the chat a couple of days ago why approach 2 would be better, @amb26 said:

> because the purpose of model state is to be a mirror for visible state
Just because you think you have certain exclusion relations in place which preclude certain combinations of such state, there's no reason use these to economise on your addressible model space ....
After all, the relations might go away some day, as a result of some further authorial decision
And then there'll you'll be, stuck with a configuration of your application which is impossible to encode
Or.... there someone will be

So it would seem that approach 2 is superior in the long term, at the cost of further development time and a bit of premature optimization. The question I have at this point to those in the reviewer role is: which approach should I take in order to get these changes merged into `master`?

BlueSlug committed:
SJRK-104: renamed func/invoker for less ambiguity with visibility states

BlueSlug committed:
SJRK-104: reintroduce `not` transform and tests for the same

BlueSlug committed:
SJRK-104: updated gpii-handlebars to latest to reduce bugginess

BlueSlug committed:
SJRK-104: edit page and editor models using relays and two values now

BlueSlug commented:
Ready for review

cindyli commented on tests/ui/js/transformsTests.js:
```diff
@@ -279,4 +279,139 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+            expectedInverseResult: false
+        },
+        "test_19": {
+            // this case does not resolve as expected. normally, !undefined === true
```
If the result of this test case is not expected, shall the transform be fixed to handle it?

cindyli commented on tests/ui/js/transformsTests.js:
```diff
@@ -279,4 +279,139 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+        }
+    };
+
+    jqUnit.test("Test negate transform function", function () {
```
Inverse results are defined but not tested.

cindyli commented on tests/ui/js/base-page-storyEditTests.js:
```diff
@@ -406,22 +406,23 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
                 {
-                    "jQueryTrigger": "click",
-                    "element": "{storyEdit}.storyEditor.dom.storyEditorNext"
+                    jQueryTrigger: "click",
```
At the change from editor to previewer and vice versa, as well as from editor step to metadata step and vice versa, please check what expected to be tracked by gpii.locationBar are tracked correctly.

BlueSlug committed:
Merge remote-tracking branch 'upstream/master' into SJRK-104

jobara committed:
SJRK-104: Add inverse transform for not transformation.

BlueSlug committed:
Merge pull request #31 from jobara/SJRK-104

SJRK-104: Add inverse transform for not transformation.

BlueSlug committed:
SJRK-104: added transform test generation function for more DRYness

BlueSlug committed:
SJRK-104: updated `undefined` transform test case comment for clarity

BlueSlug committed:
SJRK-104: added tests for historian's model state on Editor step change

BlueSlug commented:
Ready for more review

jobara commented on tests/ui/js/transformsTests.js:
```diff
@@ -253,32 +228,159 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
-            jqUnit.assertEquals("Generated string value for test case " + index + " is as expected", testCase.expectedResult, resultString);
+                // test the transformation
+                var result = fluid.model.transformWithRules(testCase, transformRules);
+                var assertionFunction = typeof testCase.expectedResult === "object" ? jqUnit.assertDeepEq : jqUnit.assertDeepEq;
```
Looks like you have a typo because both sides of the conditional point at `jqUnit.assertDeepEq`. However, the tests are passing because [`jqUnit.assertDeepEq`](https://github.com/fluid-project/infusion/blob/c3c22dc4c3a3683280989534bb5f03f8d90d1bf9/tests/test-core/jqUnit/js/jqUnit.js#L173-L179) will drop down to `jqUnit.assertEquals` if the expected or actual values are primitives. With that in mind, maybe you just call `jqUnit.assertDeepEq` always and skip the conditional check completely.

BlueSlug replied:
Nice catch! Not sure how that one slipped by. Thanks @jobara!

BlueSlug committed:
SJRK-104: fixing typo in transform tests as well as pruning a little bit

BlueSlug commented:
Ready for more review

cindyli commented on src/ui/base-page-storyEdit.js:
```diff
@@ -18,6 +18,58 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+                namespace: "manageEditorVisibility"
+            },
+            {
+                func: "{that}.events.onContextChangeRequested.fire",
```
As the change of `model.editorVisible` automatically triggers the change of `model.previewerVisible` and vice versa, which means the change to any of these 2 model values will trigger `onContextChangeRequested` event to be fired twice. Is this correct?

BlueSlug replied:
The double-firing is unintentional. I will remove the listener on `previewerVisible`

cindyli commented on src/ui/base-page-storyEdit.js:
```diff
@@ -18,6 +18,58 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
             hiddenEditorClass: "hidden"
         },
+        model: {
+            // the initial state is only the editor ui showing.
```
Could you improve this comment to better explain 3 steps for editing a story and how the visibility of them are controlled. Thanks.

cindyli commented on src/ui/base-page-storyEdit.js:
```diff
@@ -77,17 +132,22 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+                type: "gpii.locationBar",
+                options: {
+                    model: {
+                        editorVisible: "{storyEdit}.model.editorVisible",
```
Add a code comment to explain that only these 2 model values are tracked because transformations between:
1. `{storyEdit}.model.editorVisible` and `{storyEdit}.model.previewerVisible`
2. `{storyEditor}.model.editStoryStepVisible` and `storyEditor}.model.metadataStepVisible`
will set other model values properly.

cindyli replied:
bump

BlueSlug replied:
I misread your suggestion, have added a comment here :)

cindyli commented on tests/ui/js/base-page-storyEditTests.js:
```diff
@@ -408,22 +424,23 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
-                    "args": [
-                        ["{storyEdit}.storyEditor.dom.storyEditorPage1", "{storyEdit}.storyPreviewer.container"],
-                        ["{storyEdit}.storyEditor.dom.storyEditorPage2"]
+                    changeEvent: "{storyEdit}.storyEditor.applier.modelChanged",
```
Add a visibility test for clicking the "previous" button from the metadata step to the edit story step.

cindyli commented on tests/ui/js/base-page-storyEditTests.js:
```diff
@@ -747,6 +784,78 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+                    ]
+                },
+                {
+                    "jQueryTrigger": "click",
```
Please ignore my comment above. You've got the test for it here. Sorry.

BlueSlug committed:
SJRK-104: updated Edit page model comment, removed duplicate listener

BlueSlug commented:
Ready for review!

BlueSlug committed:
SJRK-104: added clarifying comment on historian model

cindyli commented:
Merged at 94a55d18c52e605dc9c7292fd0bc75b6d8887b03
